indexer/sql/src/active_model.rs

```diff
@@ -14,79 +22,263 @@ pub struct Models {
+                        tx_outcome.events.finalize.clone(),
+                        Some(transaction_id),
+                        None,
+                        created_at,
+                    )?;
+
+                    events.extend(active_models);
+                }
+            }
+        }
+
         let block = entity::blocks::ActiveModel {
             id: Set(Uuid::new_v4()),
             block_height: Set(block.info.height.try_into()?),
-            created_at: Set(naive_datetime),
+            created_at: Set(created_at),
             hash: Set(block.info.hash.to_string()),
             app_hash: Set(block_outcome.app_hash.to_string()),
         };
 
         Ok(Self {
             block,
-            ..Default::default()
+            events,
+            transactions,
+            messages,
         })
     }
 }
+
+fn flatten_events<T>(
+    block: &Block,
+    next_id: &mut EventId,
+    commitment: CommitmentStatus<T>,
+    transaction_id: Option<uuid::Uuid>,
+    message_id: Option<uuid::Uuid>,
+    created_at: NaiveDateTime,
+) -> crate::error::Result<Vec<entity::events::ActiveModel>>
+where
+    T: FlattenStatus,
+{
+    let flatten_events = flatten_commitment_status(next_id, commitment);
+    next_id.increment_idx(&flatten_events);
+
+    let mut active_models = vec![];
+    // Store previous events ids to link current event to optional parent uuid
+    let mut events_ids = HashMap::new();
+
+    for event in flatten_events {
+        let db_event_id = uuid::Uuid::new_v4();
+
+        events_ids.insert(event.id.event_index, db_event_id);
+
+        let db_event = build_event_active_model(
+            &event,
+            block,
+            transaction_id,
+            message_id,
+            db_event_id,
+            events_ids.get(&event.parent_id.event_index).cloned(),
+            created_at,
+        )?;
+
+        active_models.push(db_event);
+    }
+
+    Ok(active_models)
+}
+
+fn build_event_active_model(
+    index_event: &FlatEventInfo,
+    block: &Block,
+    tx_id: Option<uuid::Uuid>,
+    message_id: Option<uuid::Uuid>,
+    event_id: uuid::Uuid,
+    parent_event_id: Option<uuid::Uuid>,
+    created_at: NaiveDateTime,
+) -> crate::error::Result<entity::events::ActiveModel> {
+    // I'm serializing `FlattenEvent` to `serde_json::Value` and then manually
+    // removing the top hash which is serialized to.
+    // I could also use #[serde(flatten)] on `FlattenEvent`
+    let data = serde_json::to_value(&index_event.event)?;
+    // Removing the top hash
+    let data = match data {
+        Json::Object(map) => map
+            .keys()
+            .next()
+            .and_then(|key| {
+                let mut map = map.clone();
+                map.remove(key)
+            })
+            .unwrap_or_default(),
+        _ => {
+            return Err(IndexerError::Anyhow(anyhow::anyhow!(
+                "Can't get the top hash, never supposed to happen",
+            )))
```

Comment:
@penso Following the [rust style guide](https://github.com/rust-lang/api-guidelines/issues/79), don't uppercase the error message:

```diff
- Can't get the top hash, never supposed to happen
+ can't get the top hash, never supposed to happen
```

Btw if something is never supposed to happen, we can panic with `unreachable!` instead of error?